### ISSUE TITLE
Add Gurmukhi tabular figure forms and feature code support

### ISSUE THREAD

**Issue by chrissimpkins** [googlefonts/googlesans#314]

**Comment by chrissimpkins** [comment]
@kalapi will work on this during Q3 and this will be merged as part of the work for the v6.000 release

**Comment by kalapi** [comment]
Hi @MariannaPaszkowska & @chrissimpkins 

I'm attaching a new version of the Gurmukhi with tabular figures: 
- I've updated the regression test toml files to include a `tnum` test
- I've updated the browser-based tester with a visual test section for the tabular variants
- There were some inconsistencies in the heights of the default figures in `min-opsz` masters which is now fixed

Please let me know if you have any questions.

[2023-07-25_GoogleSansGurmukhi_v1.100_BETA01.zip](https://github.com/googlefonts/googlesans/files/12164022/2023-07-25_GoogleSansGurmukhi_v1.100_BETA01.zip)

P.S.: I was able to finish this quicker so finished it first :)

**Comment by madig** [comment]
@kalapi hi! Why did you change the name of `gur2_dflt_historical_below_base_consonant_forms` to include `ss01_`?

**Comment by kalapi** [comment]
Hi @madig because it was missing from the schema that we use to name lookups for fast identification.

**Comment by madig** [comment]
Our QAer found something and wanted to hear your thoughts, I shared the doc with you.

**Comment by kalapi** [comment]
@madig Can you give me access to the [latest test builds](https://drive.google.com/file/d/1-1NMpF4Sv5Iw1aH3UNDBbkE7rmhngqpb/view?usp=sharing)?

**Comment by madig** [comment]
I think I did, can you get them now?

**Comment by MariannaPaszkowska** [comment]
@madig, mind sharing the link to the doc here?

**Comment by madig** [comment]
Discussed in an internal doc

**Comment by MariannaPaszkowska** [comment]
@kalapi Could you please provide an update on the status of this issue? As highlighted by Alex in the doc above, there's an inconsistency with the positioning:

- [ ] .tf in italic appears to be off-centre, leaning slightly to the left.

- [ ] A similar discrepancy is observed in the proportional numeral set, specifically when using /bar as a reference.

Could you let us know if this is on the roadmap to be addressed? Additionally, can we have an estimated timeline for the source delivery?

**Comment by kalapi** [comment]
Hi @madig (and @MariannaPaszkowska)

I finally got a chance to look at this issue this week. I was confused with this issue because the .tf glyphs are centred in my sources while they are off-centre in the upstream sources. 

After looking into this a bit, I found that the LSB (and in-turn the RSB) values are modified when changing the internal metrics values of GlyphsApp. Our sources have different values compared to the upstream UFOs.

One way to fix this would be to update the metrics in our source, however, this seems to update the LSBs and RSBs of all glyphs which may pose a challenge of its own.

Do you have any recommendations on how I can go about fixing this?

**Comment by belluzj** [comment]
Hello @kalapi, I'm not sure what you mean by "changing the internal metrics value", but would it be possible for you to:
* create a copy of your Glyphs file
* restore the original values there, to match the UFOs; then centre the glyphs in question
* copy-paste the fixed glyphs into your original Glyphs file and give us that? (so in your original Glyphs file, only the glyphs that need fixing have been changed)

I've asked our designers if they have any Glyphs.app advice for you.

**Comment by kalapi** [comment]
> Hello @kalapi, I'm not sure what you mean by "changing the internal metrics value",[...]

These numbers control how Glyphs calculates LSB and RSB:

![Screenshot 2023-10-23 at 7 00 46 AM](https://github.com/googlefonts/googlesans/assets/15804460/8f069967-aefe-4ec7-8bef-c33bc7210fad)

> [...] but would it be possible for you to:

Yes, i'll do this now. But it'll be good to make sure that none of the other glyphs are 'moving to the left' and if they are, how it affects spacing and kerning.

Thanks for your reply.

**Comment by belluzj** [comment]
Our designers report that "We've done countless projects where we can go from italic outlines in UFO's, to a glyphsApp source, and back again, with no changes to metrics" so I'm not sure what went wrong there

**Comment by kalapi** [comment]
The issue here is not round-tripping in the same project starting from a UFO to a .glyphs and back. It's having two independent sources with different histories (and thus metadata), and when merged, takes the metadata of the upstream source. 

AFAIK this is only happening in the italic because of the italic angle and how the metrics are affected when ascender/descender values change within the source.

Here are screenshots comparing fonts generated by me vs. the test build in the max-opsz italic:

![Screenshot 2023-10-23 at 11 35 43 AM](https://github.com/googlefonts/googlesans/assets/15804460/c62cb008-34c8-4253-8cc8-2dc157b909f9)
![Screenshot 2023-10-23 at 11 35 45 AM](https://github.com/googlefonts/googlesans/assets/15804460/fed60f08-c04a-4e50-91a6-079e3a04e04d)

You can see that in the second image, the outlines are shifted to the left. It may be useful to test this in other non-Indic italics (Thai and Arabic for e.g.)

I'll perform some tests and report back. The way out maybe for me to set the metrics the same as the upstream source and adjust the x-pos of all the outlines to match the original source.

**Comment by belluzj** [comment]
Maybe the "italic sidebearings" are the wrong thing to focus on? If I understand correctly, these are generated automatically by Glyphs.app based on the italic angle + maybe the xHeight, and that would be why you see them different when your xHeight is different. The sidebearings of all the glyphs would appear different with a consistent shift to the left or the right.

The problem reported by Alex should be independent from that: he reported that /bar /digit /bar makes the digit appear as if not centred between the two bars, so it doesn't matter if the sidebearings are all a bit offset to the right or to the left, it means that there's an asymmetry between the two sides of the digits compared to a "neutral" glyph with respect to spacing (the bar). I believe that issue is fixable without shifting all other glyphs (letters/clusters etc).

![image](https://github.com/googlefonts/googlesans/assets/3616772/c80aa575-5df6-458f-bfcd-0fb865011e18)

My paint drawing to illustrate my point = it doesn't matter where the sidebearings are, the issue is the same:

![image](https://github.com/googlefonts/googlesans/assets/3616772/12b8d1fd-e403-4433-b3c2-a1966d8f746a)

Does that make sense?

**Comment by kalapi** [comment]
Hi @belluzj thanks for the explanation and illustration. I'm afraid I still don't understand what action I can take on my end to remedy this. Can you maybe make an adjustment to the `zero-gurmukhi.tf` in my Glyphs source and show me what I can do so the glyph is not off centre? In my source the .tf glyphs are as symmetrical as can be. 

I'm adding a copy of my source file here for your reference. I've added a `bar` glyph from the lated upstream UFOs and confirm that it has the same LSB/ADV/RSB as the upstream sources.

Apologies in advance for this back and forth

[GoogleSansGurmukhi-Italic.glyphs.zip](https://github.com/googlefonts/googlesans/files/13113441/GoogleSansGurmukhi-Italic.glyphs.zip)

**Comment by belluzj** [comment]
I'll have a call with a designer on our side later this afternoon to have them validate what I'm saying and apply a sample fix to the Glyphs file that you provided, thanks for sharing the file.

**Comment by joanaranito** [comment]
Hey @belluzj, just to point out that the /bar is part of the Latin set/sources, so its side bearings are not being shifted in the same way as all Gurmukhi outlines are. The bar is symmetrically spaced, whereas all Gurmukhi glyphs are shifted, so the bar is not a "neutrally-spaced" glyph for the Gurmukhi (it should be - and it would be if the Gurmukhi characters were not all being shifted)

**Comment by belluzj** [comment]
Thanks Joana, so maybe Alex's feedback is using the wrong glyph for the comparison? Or do you say it would actually be good to shift all the Gurmukhi glyphs a bit to match the rest of Google Sans?

**Comment by joanaranito** [comment]
I think Alex is using the right glyph, the problem is that the Gurmukhi is ending up with shifted side-bearings, so when combined with Latin glyphs/punctuation, it will look off-center. This is not the way we designed the Gurmukhi, it should be centered like the Latin (As Kalapi mentioned, our Gurmukhi sources are as centered as can be). Which is why we think there's something odd happening to the italic metrics when merging upstream.

**Comment by joanaranito** [comment]
Ok, Jany, I think you pointed out what's causing the issue:

"If I understand correctly, these are generated automatically by Glyphs.app based on the italic angle + maybe the xHeight, and that would be **why you see them different when your xHeight is different.** "

I just checked and the Gurmukhi sources do have a different x-height. So maybe fixing that will fix the shifted sidebearings issue upstream?

**Comment by belluzj** [comment]
I think fixing the xHeight in your sources will place the italic sidebearings in the same place as in the Latin, and then you will indeed have to shift all the outlines a little bit to match the Latin... maybe we should check with @MariannaPaszkowska if that change is necessary, or if it should only be done to the tabular figures?

**Comment by joanaranito** [comment]
I can try to do that in our sources and see if it has any effect on the Glyphs file. And if we do need to shift the outlines, maybe we can automate the shift, because it would be the same for all glyphs

**Comment by joanaranito** [comment]
Yes @belluzj, you're right, it does shift all the outlines. But shifting them back is easy to automate, and the Gurmukhi set is not big, so I'd say it's worth doing it. Kalapi is offline now, I'll catch up with him as soon as possible and we'll get back to you :)

**Comment by kalapi** [comment]
Hello @belluzj 

Apologies for the back and forth. Adding the updated package here:

[2023-10-24_GoogleSansGurmukhi_v1.100_BETA02.zip](https://github.com/googlefonts/googlesans/files/13121813/2023-10-24_GoogleSansGurmukhi_v1.100_BETA02.zip)

**Comment by kalapi** [comment]
@MariannaPaszkowska 

> Could you let us know if this is on the roadmap to be addressed? Additionally, can we have an estimated timeline for the source delivery?

I've added the updated sources above.

**Comment by belluzj** [comment]
@MariannaPaszkowska an update on this: I've imported the new sources and Alex is scheduled to review the shift tomorrow.

While regenerating the shaping tests, I noticed that the advance width of some Gurmukhi glyphs has changed by 1 or 2 units. In theory it shouldn't be necessary to change the advance width when the goal is just to shift everyone by a given amount; however I can imagine this happening with the slanted side bearings in Glyphs.app, because that computation is floating-point and needs to round down to integer, and so can introduce 1-unit differences.

What should we do about the advance width differences?

They're tiny so I would say we can leave them, since in v12 we're going to change some spacing in LGC, but there's also the option to force the advances to match those from main. We could do that in the UFOs directly to ensure no rounding from the editor is happening.
